design/KEPS/multicluster/proposal.md

```diff
@@ -0,0 +1,160 @@
+- Ambient is not going to be touched in this effort.
+- Multi-mesh support - multi-cluster is focused on single mesh support only but shouldn’t prevent Kiali from supporting multi-mesh in the future.
+
+# Solution
+
+A single Kiali configured to access each Kubernetes cluster that is part of the mesh. Kiali will require Metrics/traces to be aggregated in a centralized datastore that Kiali can query.
+
+In a multicluster deployment, Kiali will only support the OIDC auth strategy for now.
+
+## Deployment models
+
+In each of the deployment models, these are the requirements for a centralized Kiali
+
+1. Access to each kube API server that is part of the mesh
+2. Istiod access\*
+3. Aggregated metrics/traces
+4. OIDC auth strategy
+
+Regarding istiod access, if the istiod pods are running within the kube clusters that are part of the multi-cluster mesh, Kiali will access these by portforwarding using the kube API. Kiali won't support controlplanes that are completely external to the mesh but if the istiod requirement becomes optional in a future version of Kiali then Kiali could support external controlplanes as well.
+
+A service account in each cluster will provide Kiali with access to that cluster's kube API. Credentials for that service account will be stored in a secret and used by the Kiali pod to communicate with the external Kube apis.
+
+Note that a single or different network topology won't affect Kiali's configuration and deployment since Kiali will only need access to the Kubernetes API.
+
+### Multi Primary deployment
+
+![Multi-Primary](Multi-Primary.png "Multi Primary")
+
+### Primary-Remote deployment
+
+![Primary-Remote](Primary-Remote.png "Primary Remote")
+
+## Aggregated metrics and traces
+
+Kiali will require metrics/traces to be aggregated into a centralized datastore rather than aggregating this information itself. It is the user’s responsibility to ensure that metrics/traces are centralized by some means. Metrics and traces will need to include a cluster identifier as an attribute/label on each metric so that Kiali can distinguish between workloads/services running on separate clusters. [Istio standard metrics](https://istio.io/latest/docs/reference/config/metrics/#labels) includes `destination_cluster` and `source_cluster`. Kiali already uses these values for cluster boxing. Traces do no include a cluster identifier but one can be added by creating a telemetry api object. Example:
+
+```
+apiVersion: telemetry.istio.io/v1alpha1
+kind: Telemetry
+metadata:
+  name: mesh-default
+  namespace: istio-system
+spec:
+  tracing:
+  - customTags:
+      cluster:
+        literal:
+          value: Kubernetes
+```
+
+[query_scope](https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.prometheus.query_scope) can be used to limit what Kiali sees from the prometheus and jaeger queries to include only workloads/services within the mesh if these datastores are shared with non-mesh workloads.
```

Comment:
Limiting the scope of prom/jaeger will work for multiple meshes but limiting the kube objects that Kiali pulls isn't supported yet. Would it be confusing to mention multiple meshes here since that isn't part of this proposal?

Reply:
Maybe what I wrote wasn't clear.  The proposal is for Kiali to have a view into an entire [single] mesh, where that mesh may be deployed across multiple clusters.  But we are also requiring unified metrics and traces.  What I don't think we should do is require that the unified metric and/or trace DB be limited to the one mesh for that Kiali.  Rather, a user could unify metrics or traces from multiple meshes, and possibly have a Kiali for each mesh.  But those Kiali's should not query cross-mesh.  In that case the users would have to label each metric or trace with a mesh_id, and use query_scope to limit it's queries to the mesh for that Kiali instance.

Reply:
>  In that case the users would have to label each metric or trace with a mesh_id, and use query_scope to limit it's queries to the mesh for that Kiali instance.

This will work for metrics/traces where you have data from multiple meshes in your unified store but if those meshes exist on the same Kube clusters that Kiali is configured to talk to, Kiali currently has no way to distinguish between objects that belong in one mesh vs. the other and adding the support isn't part of this proposal. I was just wondering if mentioning multiple meshes would be confusing if Kiali only partially (through query scope) supports multiple meshes?

Reply:
Does this multi-cluster scenario differ from today when querying a Kube API?  We always include namespace in the query and two meshes don't share a namespace, right?  Maybe I'm getting confused.  Anyway, you can leave out multi-mesh if you like.